resources/aws_ec2.go

```diff
@@ -1031,6 +1103,92 @@ func (obj *AwsEc2Res) snsConfirmSubscription(topicArn string, token string) erro
 	return nil
 }
 
+// snsProcessEvents unmarshals instance state-change notifications and, if the
+// event matches the instance we are watching, returns an awsEc2Event.
+func (obj *AwsEc2Res) snsProcessEvent(message, instanceName string) (awsEc2Event, error) {
+	// unmarshal the message
+	var msg postMsg
+	if err := json.Unmarshal([]byte(message), &msg); err != nil {
+		return awsEc2EventNone, err
+	}
+	// check if the instance id in the message matches the name of the
+	// instance we're watching
+	diInput := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(msg.InstanceID)},
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String(instanceName)},
+			},
+		},
+	}
+	diOutput, err := obj.client.DescribeInstances(diInput)
+	if err != nil {
+		return awsEc2EventNone, err
+	}
+	// return the appropriate awsEc2Event
+	if len(diOutput.Reservations) != 0 {
+		switch msg.State {
+		case "running":
+			return awsEc2EventInstanceRunning, nil
+		case "stopped":
+			return awsEc2EventInstanceStopped, nil
+		case "terminated":
+			return awsEc2EventInstanceTerminated, nil
+		}
+	}
+	return awsEc2EventNone, nil
+}
+
+// snsAuthorize adds the necessary permission for cloudwatch to publish to the SNS topic.
+func (obj *AwsEc2Res) snsAuthorizeCloudWatch(topicArn string) error {
+	// get the topic attributes, including the security policy
+	gaInput := &sns.GetTopicAttributesInput{
+		TopicArn: aws.String(topicArn),
+	}
+	attrs, err := obj.snsClient.GetTopicAttributes(gaInput)
+	if err != nil {
+		return err
+	}
+	// get the existing security policy
+	pol := attrs.Attributes[SnsPolicy]
+	// unmarshal the current sns security policy
+	var policy snsPolicy
+	if err := json.Unmarshal([]byte(*pol), &policy); err != nil {
+		return err
+	}
+	// construct a policy statement
+	permission := snsStatement{
+		Sid:    SnsPolicySid,
+		Effect: SnsPolicyEffect,
+		Principal: snsPrincipal{
+			Service: SnsPolicyService,
+		},
+		Action:   SnsPolicyAction,
+		Resource: topicArn,
+	}
+	// add the new policy statement to the existing one
+	policy.Statement = []snsStatement{policy.Statement[0], permission}
```

Comment:
can we do:

policy.Statement = append(policy.Statement, permission) ?

AND can we check that policy.Statement isn't nil first ? Remember even if the aws API should always have a value there we should not trust that.

Reply:
@purpleidea That's what I had here before, but in the case that we don't exit cleanly and delete the topic, when we re-run SetTopicAttributes hangs indefinitely when you try to add a policy with duplicate statements. ie. []snsStatement{policy.Statement[0], permission, permission}
You're right that I need to add a check, and maybe the correct approach would be to compare any additional statements to 'permission' and if they match, return early since our job is done. What do you think?